Bug 1876515: Fix inline taskSpec error in pipeline page

karthikjeeyar opened the pull request:
Fixes: https://issues.redhat.com/browse/ODC-4724

Problem:
Issue with inline spec over taskref causes the pipeline details page to break

Solution:
Handle the inline spec issue and skip the visualization

Screenshots:
![image](https://user-images.githubusercontent.com/9964343/93078533-432a3680-f6a8-11ea-9558-8d5b415fc48e.png)

Test cases:
![image](https://user-images.githubusercontent.com/9964343/92383354-c9380180-f12b-11ea-81ab-f4affef2af7c.png)
PipelineVisualization tests
![image](https://user-images.githubusercontent.com/9964343/93078626-74a30200-f6a8-11ea-9c73-18c8a05e4d86.png)

pipeline-actions tests
![image](https://user-images.githubusercontent.com/9964343/93078752-a916be00-f6a8-11ea-892d-26ecdf010c70.png)

divyanshiGupta commented on frontend/packages/dev-console/src/utils/pipeline-utils.ts:
```diff
@@ -181,7 +181,16 @@ const appendPipelineRunStatus = (pipeline, pipelineRun) => {
 };
-
+export const hasInlineTaskSpec = (pipeline: K8sResourceKind) => {
+  const tasks = (pipeline.spec && pipeline.spec.tasks) || [];
```
```suggestion
  const tasks = pipeline.spec?.tasks ?? [];
```

divyanshiGupta commented on frontend/packages/dev-console/src/utils/pipeline-utils.ts:
```diff
@@ -192,7 +201,7 @@ export const getPipelineTasks = (
   // Each unit in 'out' array is termed as stage | out = [stage1 = [task1], stage2 = [task2,task3], stage3 = [task4]]
   const out = [];
-  if (!pipeline.spec || !pipeline.spec.tasks) {
+  if (!pipeline.spec || !pipeline.spec.tasks || hasInlineTaskSpec(pipeline)) {
```
```suggestion
  if (!pipeline.spec?.tasks || hasInlineTaskSpec(pipeline)) {
```

sahil143 reviewed (approved):
/lgtm

verified changes locally

sahil143 commented:
/lgtm cancel

It breaks in the Edit Pipeline page
<img width="574" alt="Screenshot 2020-09-09 at 3 26 47 PM" src="https://user-images.githubusercontent.com/9278015/92584142-e5b17680-f2b0-11ea-982c-c79d27720b81.png">

karthikjeeyar commented:
Added the check in edit scenario, so that if there is an inline taskSpec, it will not throw error and will allow us to add/modify the tasks in the builder form.
![image](https://user-images.githubusercontent.com/9964343/92587115-d3393c00-f2b4-11ea-8746-3322227c1c02.png)

andrewballantyne commented:
@karthikjeeyar I think we need UX here... I don't think it's appropriate to edit a Pipeline and clear out the tasks because they are using inline spec - this could have wildly unexpected results for the user. I think we need to take a path of least resistance here - it is an unsupported feature of the DevConsole. Maybe we showcase a message saying "Using tasks inlineSpec, which is unsupported" and removing the Edit Pipeline kebab option. 

cc @bgliwa01 / @openshift/team-devconsole-ux

andrewballantyne commented on frontend/packages/dev-console/src/utils/pipeline-utils.ts:
```diff
@@ -181,7 +181,16 @@ const appendPipelineRunStatus = (pipeline, pipelineRun) => {
-
+export const hasInlineTaskSpec = (pipeline: K8sResourceKind) => {
+  const tasks = pipeline?.spec.tasks ?? [];
+  let hasInlineSpec = false;
+  tasks.forEach((task) => {
+    if (!hasInlineSpec) {
+      hasInlineSpec = !!(task.taskSpec && !task.taskRef);
+    }
+  });
```
You're re-implementing `Array.prototype.some`... best just [use it instead](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some).

karthikjeeyar replied:
Updated it.

bgliwa01 commented:
@karthikjeeyar I agree with @andrewballantyne we should remove the edit pipeline option in the kebab dropdown and then we can display the message "This Pipeline cannot be visualized. Pipeline taskSpec is not supported."

andrewballantyne commented:
> @karthikjeeyar I agree with @andrewballantyne we should remove the edit pipeline option in the kebab dropdown and then we can display the message "This Pipeline cannot be visualized. Pipeline taskSpec is not supported."

@karthikjeeyar When you do this, make sure there is a gap under the message before the first label.

Thanks @bgliwa01 !

karthikjeeyar commented:
Updated the code to hide the edit pipeline option and added the UX provided message to the visualization area.
 @bgliwa01 Please take a look at the updated screenshots.
cc: @andrewballantyne

andrewballantyne commented on frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx:
```diff
@@ -17,21 +18,32 @@ const PipelineVisualization: React.FC<PipelineTopologyVisualizationProps> = ({
+  } else if (nodes.length === 0 && edges.length === 0) {
     // Nothing to render
     // TODO: Confirm wording with UX; ODC-1860
-    return <Alert variant="info" isInline title="This Pipeline has no tasks to visualize." />;
```
Why did you remove the return? I think we can just have another return here.

karthikjeeyar replied:
I have moved it inside the block, so that it gets the margin-bottom style (for the gap underneath the message) from the `odc-pipeline-visualization` class.

andrewballantyne replied:
Ah, I see... why don't we clean this up more and instead of treating errors differently, just have a variable called `content` or something and have 3 if-else statements... landing on rendering the Component as the fallback `else`.

It'll look cleaner and be easier to read.

karthikjeeyar replied:
I agree, its more readable now. Updated it

andrewballantyne commented on frontend/packages/dev-console/src/utils/pipeline-utils.ts:
```diff
@@ -181,7 +181,10 @@ const appendPipelineRunStatus = (pipeline, pipelineRun) => {
 };
-
+export const hasInlineTaskSpec = (pipeline: K8sResourceKind) => {
+  const tasks = pipeline?.spec.tasks ?? [];
```
`spec` is not optional.

andrewballantyne commented on frontend/packages/dev-console/src/utils/pipeline-utils.ts:
```diff
@@ -192,7 +195,7 @@ export const getPipelineTasks = (
   // Each unit in 'out' array is termed as stage | out = [stage1 = [task1], stage2 = [task2,task3], stage3 = [task4]]
   const out = [];
-  if (!pipeline.spec || !pipeline.spec.tasks) {
+  if (!pipeline?.spec.tasks) {
```
Don't think this can ever happen anymore. Tasks of 1 is required by the operator. We can keep this if you change it to look for tasks also being an empty array. 

Also `spec` is not optional.

andrewballantyne commented on frontend/packages/dev-console/src/utils/__tests__/pipeline-test-data.ts:
```diff
@@ -12,7 +12,7 @@ export const mockPipelinesJSON: K8sResourceKind[] = [
       params: [
         {
           name: 'APP_NAME',
-          type: 'string',
```
`type` should be allowed 🤔  Why did you remove this?

andrewballantyne replied:
Ah because it's not part of the type 😞 

Can you just add it to the type instead? 
It's [supported by upstream](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#specifying-parameters)
> Each declared parameter has a type field, which can be set to either array or string.

We'll want the type to probably be:

```
  type: 'string' | 'array';
```
Instead of a straight `string` type... this way we can easily see what we support.

bgliwa01 reviewed (approved):
lgtm

jerolimov commented on frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.tsx:
```diff
@@ -33,10 +33,8 @@ interface PipelineVisualizationTaskProp {
-      kind?: string;
-    };
+    taskSpec?: PipelineTaskSpec;
+    taskRef?: PipelineTaskRef;
```
:+1:

jerolimov reviewed (approved):
Tested it locally, works fine :+1: 

/lgtm

jerolimov commented:
/retest

andrewballantyne commented:
/cherry-pick release-4.5
